### PR TITLE
feat: Major compaction feature

### PR DESCRIPTION
**Pull request by wayblink**
#32939

### CODE REVIEW

File: pkg/util/typeutil/schema.go [hunk 1]
```diff
@@ -377,6 +377,20 @@ func IsDenseFloatVectorType(dataType schemapb.DataType) bool {
 	}
 }
 
+// return VectorTypeSize for each dim (byte)
+func VectorTypeSize(dataType schemapb.DataType) float64 {
+	switch dataType {
+	case schemapb.DataType_FloatVector, schemapb.DataType_SparseFloatVector:
+		return 4.0
+	case schemapb.DataType_BinaryVector:
+		return 0.25
```
Comment (czs007) [inline pkg/util/typeutil/schema.go:386]: 1/8?

File: pkg/util/paramtable/component_param.go [hunk 2]
```diff
@@ -732,6 +735,22 @@ like the old password verification when updating the credential`,
 		Doc:          "panic or not when plugin fail to init",
 	}
 	p.PanicWhenPluginFail.Init(base.mgr)
+
+	p.UsePartitionKeyAsClusteringKey = ParamItem{
+		Key:          "common.usePartitionKeyAsClusteringKey",
+		Version:      "2.4.0",
```
Comment (czs007) [inline pkg/util/paramtable/component_param.go:741]: 2.4.2

File: pkg/util/paramtable/component_param.go [hunk 3]
```diff
@@ -732,6 +735,22 @@ like the old password verification when updating the credential`,
...
+		Doc:          "if true, do clustering compaction and segment prune on partition key field",
+		DefaultValue: "false",
+	}
+	p.UsePartitionKeyAsClusteringKey.Init(base.mgr)
+
+	p.UseVectorAsClusteringKey = ParamItem{
+		Key:          "common.useVectorAsClusteringKey",
+		Version:      "2.4.0",
```
Comment (czs007) [inline pkg/util/paramtable/component_param.go:749]: 2.4.2

File: pkg/util/paramtable/component_param.go [hunk 4]
```diff
@@ -3065,6 +3102,137 @@ During compaction, the size of segment # of rows is able to exceed segment max #
...
+		DefaultValue: "false",
+		Doc:          "Enable clustering compaction",
+		Export:       true,
+	}
+	p.ClusteringCompactionEnable.Init(base.mgr)
+
+	p.ClusteringCompactionAutoEnable = ParamItem{
+		Key:          "dataCoord.compaction.clustering.autoEnable",
+		Version:      "2.4.0",
```
Comment (czs007) [inline pkg/util/paramtable/component_param.go:3116]: 2.4.2

File: pkg/util/merr/errors.go [hunk 5]
```diff
@@ -175,6 +176,11 @@ var (
 	ErrClusteringCompactionClusterNotSupport      = newMilvusError("milvus cluster not support clustering compaction", 2301, false)
 	ErrClusteringCompactionCollectionNotSupport   = newMilvusError("collection not support clustering compaction", 2302, false)
 	ErrClusteringCompactionCollectionIsCompacting = newMilvusError("collection is compacting", 2303, false)
+	ErrClusteringCompactionNotSupportVector       = newMilvusError("vector field clustering compaction is not supported", 2304, false)
+	ErrClusteringCompactionSubmitTaskFail         = newMilvusError("fail to sumbit task", 2305, true)
+	ErrClusteringCompactionCompactionTaskLost     = newMilvusError("compaction task lost", 2306, true)
+	ErrClusteringCompactionGetCollectionFail      = newMilvusError("fail to get collection", 2307, true)
```
Comment (czs007) [inline pkg/util/merr/errors.go:182]: fail to get collection in compaction?

File: pkg/util/merr/errors.go [hunk 6]
```diff
@@ -157,8 +157,9 @@ var (
 	ErrInvalidStreamObj     = newMilvusError("invalid stream object", 1903, false)
 
 	// Segcore related
-	ErrSegcore            = newMilvusError("segcore error", 2000, false)
-	ErrSegcoreUnsupported = newMilvusError("segcore unsupported error", 2001, false)
+	ErrSegcore             = newMilvusError("segcore error", 2000, false)
+	ErrSegcoreUnsupported  = newMilvusError("segcore unsupported error", 2001, false)
+	ErrSegcoreFakeFinished = newMilvusError("segcore fake finished", 2002, false)
```
Comment (czs007) [inline pkg/util/merr/errors.go:162]: segcore pretend finished?

File: internal/indexnode/meta.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package indexnode
```
Comment (czs007) [inline internal/indexnode/meta.go:1]: delete this file ? just one line

File: internal/proto/data_coord.proto [hunk 8]
```diff
@@ -828,3 +842,52 @@ message GcControlRequest {
...
+  int64 collectionID = 13;
+  int64 partitionID = 14;
+  string channel = 15;
+  schema.FieldSchema clustering_key_field = 16;
+  int64 max_segment_rows = 17;
+  int64 prefer_segment_rows = 18;
+  repeated int64 inputSegments = 19;
+  repeated int64 resultSegments = 20;
+  int64 analyzeTaskID = 21;
+  int64 analyzeVersionID = 22;
```
Comment (czs007) [inline internal/proto/data_coord.proto:883]: rename to analyzeVersion?

File: internal/proto/data_coord.proto [hunk 9]
```diff
@@ -828,3 +842,52 @@ message GcControlRequest {
...
+}
+
+message PartitionStatsInfo {
+  int64 collectionID = 1;
+  int64 partitionID = 2;
+  string vChannel = 3;
+  int64 version = 4;
+  repeated int64 segmentIDs = 5;
+  int64 analyzeTaskID = 6;
```
Comment (czs007) [inline internal/proto/data_coord.proto:892]: remove this. unnecessary

File: internal/proto/data_coord.proto [hunk 10]
```diff
@@ -828,3 +842,52 @@ message GcControlRequest {
...
+  timeout = 5;
+  init = 6;
+  analyzing = 7;
+  analyzed = 8;
+  indexing = 9;
+  indexed = 10;
+  cleaned = 11;
+}
+
+message CompactionTask {
```
Comment (bigsheeper) [inline internal/proto/data_coord.proto:861]: maybe `ClusteringCompactionTask` is better

File: internal/datacoord/clustering_compaction_manager.go [hunk 11]
```diff
@@ -0,0 +1,809 @@
...
+			return merr.WrapErrClusteringCompactionCompactionTaskLost(oldPlanID)
+		}
+		segmentIDs := make([]int64, 0)
+		for _, seg := range compactionTask.result.Segments {
+			segmentIDs = append(segmentIDs, seg.GetSegmentID())
+		}
+		task.ResultSegments = segmentIDs
+	}
+
+	return t.processIndexingTask(task)
```
Comment (bigsheeper) [inline internal/datacoord/clustering_compaction_manager.go:433]: so I think maybe we can just remove Compacted state?

File: internal/datacoord/clustering_compaction_manager.go [hunk 12]
```diff
@@ -0,0 +1,809 @@
...
+		allocator:         allocator,
+		compactionHandler: compactionHandler,
+		analyzeScheduler:  analyzeScheduler,
+		handler:           handler,
+	}
+}
+
+func (t *ClusteringCompactionManager) start() {
+	triggers := t.meta.GetClusteringCompactionTasks()
+	for _, tasks := range triggers {
```
Comment (czs007) [inline internal/datacoord/clustering_compaction_manager.go:95]: This logic should not be included here; it should be handled within the state machine.

File: internal/proto/data_coord.proto (hunk 10 above)
Comment (wayblink) [reply internal/proto/data_coord.proto:861]: This is for future. If we store a meta for all types of compaction.

File: internal/datacoord/clustering_compaction_manager.go (hunk 11 above)
Comment (wayblink) [reply internal/datacoord/clustering_compaction_manager.go:433]: This is to be compatible with old compaction state. Actually, yes, it compacted and indexing are the same actually. No harm to keep it, right?

File: internal/proto/data_coord.proto [hunk 13]
```diff
@@ -828,3 +842,52 @@ message GcControlRequest {
...
+  schema.FieldSchema clustering_key_field = 16;
+  int64 max_segment_rows = 17;
+  int64 prefer_segment_rows = 18;
+  repeated int64 inputSegments = 19;
+  repeated int64 resultSegments = 20;
+  int64 analyzeTaskID = 21;
+  int64 analyzeVersionID = 22;
+}
+
+message PartitionStatsInfo {
```
Comment (bigsheeper) [inline internal/proto/data_coord.proto:886]: How about `ClusteringInfo` ?

File: internal/datacoord/compaction_trigger.go [hunk 14]
```diff
@@ -249,7 +294,35 @@ func (t *compactionTrigger) triggerCompaction() error {
 	return nil
 }
 
-// triggerSingleCompaction triger a compaction bundled with collection-partition-channel-segment
+// triggerClusteringCompaction trigger clustering compaction.
+func (t *compactionTrigger) triggerClusteringCompaction() error {
```
Comment (bigsheeper) [inline internal/datacoord/compaction_trigger.go:298]: I think we need to make sure mutual exclusion between clustering compaction and L0/mix compaction. You can integrate this logic in compaction_scheduler.go.

**Comment by bigsheeper** [comment]
Currently L0 compaction only compacts L0 segments with L1 segments, please make sure L2 segments can also participate in L0 compaction:
image

**Comment by czs007** [comment]
duplicated with https://github.com/milvus-io/milvus/pull/33620

**Pull request closed without merging**